tools/buildgen/extract_metadata_from_bazel_xml.py

```diff
@@ -245,98 +242,171 @@ def _external_dep_name_from_bazel_dependency(bazel_dep):
-    bazel_label_to_dep_name = {}
-    for dep_name in public_dep_names:
-        bazel_label_to_dep_name[_get_bazel_label(dep_name)] = dep_name
-
-    target_name = target_dict['name']
-    bazel_deps = target_dict['_DEPS_BAZEL']
-
-    # initial values
-    public_headers = set(target_dict['_PUBLIC_HEADERS_BAZEL'])
-    headers = set(target_dict['_HEADERS_BAZEL'])
-    src = set(target_dict['_SRC_BAZEL'])
-    deps = set()
+def _deduplicate_append(target: List[str], pending: List[str]) -> None:
+    """Append the list without duplicated items in place.
+    
+    Expected to be used to create a order-stable dependency list.
+    """
+    seen = set(target)
+    for dep in pending:
+        if dep not in seen:
+            target.append(dep)
+
+
+def _compute_transitive_metadata(
+        rule_name: str, bazel_rules: Any,
+        bazel_label_to_dep_name: Dict[str, str]) -> None:
+    """Computes the final build metadata for Bazel target with rule_name.
+
+    The dependencies that will appear on the deps list are:
+
+    * Public build targets including binaries and tests;
+    * External targets, like absl, re2.
+
+    All other intermediate dependencies will be merged, which means their
+    source file, headers, etc. will be collected into one build target. This
+    step of processing will greatly reduce the complexity of the generated
+    CMake build files.
+
+    The final build metadata are:
+    * _TRANSITIVE_DEPS: all the transitive dependencies including intermediate
+                        targets;
+    * _COLLAPSED_DEPS:  dependencies that fits our requirement above, and it
+                        will remove duplicated items and produce the shortest
+                        possible dependency list;
+    * _COLLAPSED_SRCS:  the merged source files;
+    * _COLLAPSED_PUBLIC_HEADERS: the merged public headers;
+    * _COLLAPSED_HEADERS: the merged non-public headers.
+    * _EXCLUDE_DEPS: the intermediate dependency labels.
+
+    For the new collapsed_deps, the new algorithm improved cases like:
+
+    The previous result:
+        end2end_tests -> [grpc_test_util, grpc, gpr, address_sorting, upb]
+        grpc_test_util -> [grpc, gpr, address_sorting, upb, ...]
+        grpc -> [gpr, address_sorting, upb, ...]
+    
+    The new result:
+        end2end_tests -> [grpc_test_util]
+        grpc_test_util -> [grpc]
+        grpc -> [gpr, address_sorting, upb, ...]
+    """
+    bazel_rule = bazel_rules[rule_name]
+    direct_deps = _extract_deps(bazel_rule, bazel_rules)
+    transitive_deps = []
+    # We care about the order of dependencies list, so it will be a list.
+    collapsed_deps = []
+    # We don't care about the order of srcs and headers, they will be sorted.
+    collapsed_srcs = set(_extract_sources(bazel_rule))
+    collapsed_public_headers = set(_extract_public_headers(bazel_rule))
+    collapsed_headers = set(_extract_nonpublic_headers(bazel_rule))
+    exclude_deps = set()
```

Comment:
I experimented a bit myself and I take back what I said about Scenario 2. Due to the way our bazel BUILD is organized (e.g. grpc depends on some of the same intermediate libraries as grpc++),  implementing Scenario 2 behavior as I described it would actually lead to a broken sanity test as some .cc files would be included in our build multiple times.